### DISCUSSION TITLE
Different behaviour between using GPIO 1,3 or 0,9 with I2C

### DISCUSSION THREAD

**Discussion by hjboven** [arendst/Tasmota#14936] [Q&A] [answered]
### Problem:
I'm writing this hoping someone can replicate so we can confirm this to be a bug. Recently I upgraded my Sonoff Dual R2 from Tasmota v8.1.0 to v10.1.0  and started to experience Problems using my BME280 Sensor over i2c.

Previously I was running folowing setup. 
1. Tasmota v8.1.0 Sensors
2. Sonoff Dual R2
3. Two Buttons on GPIO 0 & GPIO 9
4. BME280 i2C Sensor on GPIO 1 & GPIO 3
The Device is used in my garden wintershelter to keep the plants from freezing. Using 2 rules to switch Heater when < 0.8 °C or to switch Ventilation > 15°C. It had been running fine for 2 Winters at that time.

I updated to v10.1.0 Sensors and had strange problems. It woud work and fail intermittently at unregular intervals. The shown symptom beeing displayed Temp of -146 °C.  It can get cold here but not that Cold :-)

### Research:
First I tought the BME280 to be defective but after some tinkering I found that when using GPIO 0 & 9 for i2c it worked fine.  getting curious I took the oscilloscope and had a look.

**Below are GPIO 0 & GPIO 9 set as i2c connected to BME280**
As you can see a reasonable square pulse and the i2c communication works fine.
![i2c-on-GPIO-0_9](https://user-images.githubusercontent.com/38787048/154985433-c59d434a-0ff1-46d4-98ca-6588d6f4a61b.png)

**Below are GPIO 1 & GPIO 3 set as i2c connected to BME280**
As you can clearly see here a very poor pulse almost sawtooth like. We are clearly missing some Pull Up Resitors here.
![i2c-on-GPIO-1_3](https://user-images.githubusercontent.com/38787048/154985717-9bfd6c89-af44-4382-a483-693894eab733.png)

**Adding two 5k6 Pull Up resistors on GPIO 1 & GPIO 3 gets it working again.** 
![i2c-on-GPIO-1_3-PullUp](https://user-images.githubusercontent.com/38787048/154986716-009dcf90-01ff-4169-86ef-f19eaa48b9ca.png)

### Question: 
As this was working for 2 Winters already without any Problems using  i2c on GPIO 1 & GPIO 3 why dit that change afther upgrading the software? Could it be that somewere down the road between  v8.1.0 and v10.1.0 we stopped setting the internal Pull Up Resistors on GPIO 1 & GPIO 3 when these are configured to i2c ??

I Foud my solution using 2 small Pull Up Resitors. Yet if it is trough that the internal Pull Up resitors are not configured we should / clould change that and avoid running into this altogether.

Regards
Henk

**Comment by arendst** [comment, marked as answer]
Thx for your thorough investigation!

As you use the default serial interface for I2C you might expect some changes from version v8.1.0 (old) to v10.1.0 where the Arduino core version has changed too from v2.6.1 to v2.7.4.9.

So it might well be the internal pull-ups have been disabled in between the versions to better account for a more solid I2C connection were the connected I2C device is equiped with the correct termination (pull-ups).

**Comment by arendst** [comment]
Looking at the I2C core code I see there are still pullups enabled:
```
void Twi::init(unsigned char sda, unsigned char scl)
{
    // set timer function
    ets_timer_setfn(&timer, onTimer, NULL);

    // create event task
    ets_task(eventTask, EVENTTASK_QUEUE_PRIO, eventTaskQueue, EVENTTASK_QUEUE_SIZE);

    twi_sda = sda;
    twi_scl = scl;
    pinMode(twi_sda, INPUT_PULLUP);
    pinMode(twi_scl, INPUT_PULLUP);
    twi_setClock(preferred_si2c_clock);
    twi_setClockStretchLimit(150000L); // default value is 150 mS
}
```

In that case it might well be you experience power issues we see on many Sonoff devices when newer tasmota versions are used with default more features enabled. You might want to chk the voltage between the two versions.

**Reply by hjboven** [reply]
When adding external PulUp resistors is curing the Problem it leaves me struggling to believe Powerstability to be the issue. In my understanding I'm using more not less power with the external resistors. Then again I've learned I(O)T not to be an exact science and I will measure VCC and look for transients on the v10.1.0 version as soon i've finished my evening beer.

**Reply by hjboven** [reply]
As I wrote the first Posting I left the Sonnoff with the working BME280 wired on GPIO 0 & GPIO 9 (Internal PullUps work fine on those). 

VCC at the BME280 is stable at 3.294 V, and the osciloscope sees insignificant digital noise on it.
Yet as I looked to SDA & SCL on the osciloscope I saw i2c stopped decoding again eventough the waveform is fine.  A brief look at the web interface shows that all to familiar -146 °C again. 

I'm slowly getting towards suspecting that we might even have 2 Problems:

1. Using the i2c on GPIO 1 & GPIO 3 definitly does not activate the internal PullUps or their resistance is to high. Wen can cure that using two external PullUps. 4k7 to 5k6 seems to work nicely.

2. i2c with my BME280's seems prone to hangups after some time even when using GPIO 0 & GPIO 9. I've even become to believe I might be chasing a software problem As I can replicate this behaviour with 2 diferent Sonoffs both using v10.1.0 Sensor, and using 4 diferent BME280's. A test using a BM**P**280 on the same sonoff's using the same GPIO's was rocksolid > 24h. 

Maybe I'll try and replicate this on a D1 mini tomorrow but for now I've done enough Problem chasing. 

It would be good to know if someone els experiences the same problems wit BME280 under Tasmota v10.1.0 Sensor. Just to rule out that, god forbid, all my BME280's to have the same problem.

**Comment by Jason2866** [comment]
Verified Sonoff Dual you use is a ESP8285? If it is a ESP8266 you can not use GPIO 9. If it worked before you have been lucky. GPIO 9 is used for connecting the flash chip to the esp8266

**Reply by sfromis** [reply]
ESP8285 should be what is used. https://github.com/arendst/Tasmota/issues/1249

**Reply by hjboven** [reply]
Hi Jason,

The Sonoff Dual R2 uses an ESP8285. And yes it worked fine before.
GPIO 0 & 9 are the two Button Contacts on the Printheader.
GPIO 1 & 3 are the TX and RX on the other Printheader.
![Dual-R2-2](https://user-images.githubusercontent.com/38787048/155168083-2c743006-096b-49f7-ad69-073cf64c2478.jpg)

The main Problem remains that the GPIO 1 & 3 are not activating the internal PullUps when configured as SDA & SCL.

I cannot replicate on other devices as I have no other ESP8255's than those two Sonnoff Dual R2

**Comment by Jason2866** [comment]
To answer your question all my BME280 and BMP280 run without any issue with latest dev version.
None is connected to a sonoff device (NodeMCU and D1 mini where VDO is replaced with a stronger one)..
